### PR TITLE
Web onboarding: Add SAML applications to resource checking

### LINKED ISSUE #31556: Web Onboarding: Add saml applications for resource checking
The screenshot below shows me the `add your first resource` onboard dialogue, but in the background i have a SAML app. This onboarding dialogue should only render if there are 0 resources added.

We need to add `saml application` kind when [resource checking](https://github.com/gravitational/teleport/blob/abde17e4ce848d38dcba46deab8cec2bb68ea583/lib/web/resources.go#L49)

image

### PR DESCRIPTION
**Pull request by kiosion**
Resolves #31556 

## Changes
* Use `clusterUnifiedResourcesGet` instead of `checkAccessToRegisteredResource` in frontend so resource check includes SAML apps, and doesn't erroneously show onboarding popup on login if only SAML apps are accessible to user.
* Include all resources (accessible+request-able) so users with only request-able resources and permission to enrol new resources don't see onboarding popup on login.
* Add todos to remove `checkAccessToRegisteredResource` post-18.0 as it's not used elsewhere.

### CODE REVIEW

File: lib/web/resources.go
Comment (kiosion) [inline lib/web/resources.go:53]: Is this all that's needed? It works from my testing but I could be missing something.

Comment (zmb3) [reply lib/web/resources.go:53]: This works, but you might want to convert this to use `ListUnifiedResources` so that you can make one API call (instead of 6 in the worst case).

Comment (flyinghermit) [reply lib/web/resources.go:53]: Seems like that might do the trick.

Though, while we are at it, might I suggest couple things?

1. Can we change the name of this method? Because it says `checkAccessToRegisteredResource ` but inside the method we return `HasResource: true` even if user does not has access to resource.

```
res, err := clt.ListResources(...)
if err != nil {
	if trace.IsAccessDenied(err) {
		continue
	}
	return nil, trace.Wrap(err)
}

if len(res.Resources) > 0 {
	return checkAccessToRegisteredResourceResponse{
		HasResource: true,
	}, nil
}
```

Maybe `checkClusterResourceCount` or something similar?

2. Can we move the Web UI logic somewhere in the Unified resource view?

Currently, it is called inside the `TeleportContext` which is initialized very early in the UI bootstrap process. So even if there are resources, we will be needlessly firing the `resources/check` API the first time user signs in to Teleport (given enough privilege to add new resource as [checked here](https://github.com/gravitational/teleport/blob/8fb974f464bdc952da680cb4e0de17d0ea0328d6/web/packages/teleport/src/teleportContext.tsx#L101-L104)).

Comment (flyinghermit) [reply lib/web/resources.go:53]: I am not sure what changes would the 2nd suggestion entail, so given the scope of your PR, feel free to tackle it later if you wish to :)

Comment (kiosion) [reply lib/web/resources.go:53]: Gotcha, makes sense to me – the func name did confuse me at first. I'll look into using `ListUnifiedResources` here, & work on those other changes in another PR

**Review by zmb3 — APPROVED** [review]
Looks a lot cleaner and is probably more efficient too.

Note: I did not test this, but the code looks good assuming it works.

File: lib/web/resources.go
Comment (flyinghermit) [inline lib/web/resources.go:67]: tbh, I am still unclear on the intended behavior here:
1. Show the onboarding if the user has permission to create a new resource and there is exactly `zero` resource in cluster. OR 
2. Show the onboarding  if the user has permission to create a new resource but we could not fetch any resource accessible by that user. (could be due to `zero` resource in cluster or due to missing permission).

I checked this with a proper editor role that can create any type of new resource, but without the wildcard labels against a cluster that has resource configured with labels, the modal still pops up. Meaning, for users who has permission to create any type of resource but lacks access to existing resource gated behind specific labels, we will still show them the onboarding modal (the 2nd case)

Is that the behavior we want here?

Comment (kiosion) [reply lib/web/resources.go:67]: I believe so? # 2 makes sense to me, but I don't have much more ctx about this modal besides the fact it should also take SAML apps into account.

Comment (flyinghermit) [reply lib/web/resources.go:67]: Not a huge deal either way, I just wanted to confirm if we are not changing originally intended behavior here.  

Looking at the [original ticket](https://github.com/gravitational/teleport/issues/13997) for which this was implemented. Under **Auto Show Wizard** section, it says to show when
- User has the right perms
- Querying for each agent has no result

I think we wanted to implement Option 1? 

@kimlisa Can you please confirm the behavior, when do we want to show the onboarding modal?

Comment (kimlisa) [reply lib/web/resources.go:73]: the way it currently works now was probably intended, I listed it as a requirement here: https://github.com/gravitational/webapps/pull/1121, which tells me that we were aware of it

File: lib/web/resources.go
Comment (avatus) [reply lib/web/resources.go:65]: This is a quirk of the frontend/backend storage of the unified resources. the frontend uses `KindApp` but the unified resources cache is stored as `KindAppServer` (but unique by contained `App.GetName()`). This request essentially makes search through the unified cache based on a name index, so only the "contained" type matters.

I'd take a look at [how we make a unified requests from the frontend](https://github.com/gravitational/teleport/blob/master/lib/web/apiserver.go#L2733) to determine the use of the Kinds field, but also, note that we store them by these types

```go
// UnifiedResourceKinds is a list of all kinds that are stored in the unified resource cache.
var UnifiedResourceKinds []string = []string{
	types.KindNode,
	types.KindKubeServer,
	types.KindDatabaseServer,
	types.KindAppServer,
	types.KindWindowsDesktop,
	types.KindSAMLIdPServiceProvider,
}
```

This is because, when we display an app or database on the frontend, we are only displaying information about the "contained" app, therefore when querying or pinning or whatever, its done by the name visible to the users (and we don't display app servers on the frontend, only apps). 

Does that help?

File: lib/web/resources.go
Comment (flyinghermit) [reply lib/web/resources.go:73]: Thanks for confirming. 

It's option 2 then. In that case, can we utilize existing `clusterUnifiedResourcesGet` method here which already checks for `KindSAMLIdPServiceProvider` resource and handles all the other edge cases?

(I personally am in the option 1 bandwagon)

Comment (avatus) [reply lib/web/resources.go:73]: Option 1 sounds great to me as well

File: lib/web/resources.go
Comment (kiosion) [reply lib/web/resources.go:65]: yes, thanks for the clarification!

Comment (kimlisa) [reply lib/web/resources.go:65]: i'm probably really confused, and correct me if i'm wrong, but we shouldn't need the following: `KindApp, KindDatabase, KindDatabaseService, KindWindowsDesktopService, KindKubernetesCluster`. 

and we also don't need these when we construct the `default` kinds when listing all resources?

read this to see an example explanation for [db (KindDatabase) vs db_server (KindDatabaseServer) vs db_service (KindDatabaseService)](https://gravitational.slack.com/archives/C01TYKHFVTQ/p1692868723813079?thread_ts=1692821967.429929&cid=C01TYKHFVTQ)

`listing unified resources` (i thought) was to list resources that users can potentially connect to. this is how `tsh db ls` and `tctl db ls` works, it's essentially get `db_servers`.

to say "give me a list of resources of `types.Database`" means to give me a resource of all `kind: db` regardless if there is a `db_server` for it or not. a standalone `db` can exist if there are no `db_service` to pick it up.

UNLESS... that's what we wanted to do? Like is list unified resources just a way to get a list of every resource that's supported? (and i'm also unsure if the same pattern holds with desktops and kubes)

also, i _think_ we may have a bug [here](https://github.com/gravitational/teleport/blob/d2cf47f7d352edea1462baf8233fde5c4ed5643e/lib/auth/auth_with_roles.go#L1393), i dont think we are perm checking for kinds that are not in this [list](https://github.com/gravitational/teleport/blob/d2cf47f7d352edea1462baf8233fde5c4ed5643e/lib/services/unified_resource.go#L43), like the lookup can fail, but we seem to ignore it?

Comment (avatus) [reply lib/web/resources.go:65]: We ignore the RBAC errors if they [are equal the amount of kinds they requested](https://github.com/gravitational/teleport/blob/d2cf47f7d352edea1462baf8233fde5c4ed5643e/lib/auth/auth_with_roles.go#L1398). If the callsite (the UI in this case) calls every kind, but they only have access to Servers and Kubes, we don't want to show an RBAC error for "no access to DBs", we just will show them the results of the kinds _they do_ have access to.

We set defaults at the api layer level (and at the tsh level in Connect), but not at the request level. The reason is that we can evolve the apiserver with more types, but only return the types the client requests in order to not return types it cant understand (for example, the web UI would break if it received a type it didn't know how to display).

We wanted to make the kind explicitly set to ensure a client would only request what it can access, but in this case we are requesting to get a count so all available types should probably be added. Maybe a default deeper in would make sense for this case, and then elsewhere the client code would still require it be set.

Comment (kimlisa) [reply lib/web/resources.go:65]: > We ignore the RBAC errors if they [are equal the amount of kinds they requested](https://github.com/gravitational/teleport/blob/d2cf47f7d352edea1462baf8233fde5c4ed5643e/lib/auth/auth_with_roles.go#L1398). If the callsite (the UI in this case) calls every kind, but they only have access to Servers and Kubes, we don't want to show an RBAC error for "no access to DBs", we just will show them the results of the kinds they do have access to.

sorry, i meant if we request for kinds like `types.KindApp` for example, since this type is not in this [list](https://github.com/gravitational/teleport/blob/d2cf47f7d352edea1462baf8233fde5c4ed5643e/lib/services/unified_resource.go#L43) we use to make a lookup of access errors, it doesn't get counted as an error (if there is an access error), so if we have a request where kinds == `types.KindApp, types.KindDatabase`, and user RBAC denies both, it won't return an error (but should according to code). You can probably request kind of any string value and it's error will also be ignored (which maybe means we should check that the requested Kind is supported or not?)

for the default values, my question was more like, what will be the difference between (look at comment)

```go
	// set default kinds to be requested if none exist in the request
	if len(kinds) == 0 {
		kinds = []string{
			types.KindApp, // change to types.KindAppServer
			types.KindDatabase, // change to types.KindDatabaseServer
			types.KindNode,
			types.KindWindowsDesktop,
			types.KindKubernetesCluster,
			types.KindSAMLIdPServiceProvider,
		}
	}
```

but i did a little testing, and this is where my confusion lies.
- i had some `types.KindAppServer` resources: requesting with `types.KindAppServer` returned me no results (which is probably what @kiosion experienced right?), but  requesting with `types.KindApp` returned me the result (which is confusing b/c I do have `app_servers`)
- similarly, i had some standalone `types.KindDatabase` resources (without a db_server), and requesting with `types.KindDatabase` returned me no result, but the way we use the api here make it look like i should be getting some results b/c i do have `db`s

i hope that made sense, it's opposite to how `ListResources` work and how listing works with our CLI tools, so the api is confusing. and idk if this was intended or we got confused along the way (which is understandable lol)

anyways, now that I understand this quirk, @kiosion the list should be identical with the `default` we set when we list unified resources, i'd make a const of it

Comment (avatus) [reply lib/web/resources.go:65]: The reason `AppServer` doesn't get returned in the request is because [we check if the contained resource type](https://github.com/gravitational/teleport/blob/master/api/types/resource.go#L506) exists in the `kinds` field during our filtering. [this is where the conversion happens](https://github.com/gravitational/teleport/blob/master/lib/services/matchers.go#L175). Because the unified resource cache stores an item by `name-type`, the frontend will only know of the `KindApp`, which means we have to store things by the contained app, otherwise the frontend and unified cache would show different results. 

The web UI is only concerned with `KindApp` and not `KindAppServer` because we could only have 1 app but 10 app servers that serve that app. Also, I was under the impression that these are 1-to-1 right now (not in uniqueness but just in terms of functionality, it turned out that 1 app server serves 1 app) and I believe it's mandatory to have? I could be wrong but I think i remember that somewhere. This is why the request from the _client_ asks about the types it knows how to display, the server converts and interprets that to what apps they are actually trying to retrieve from the unified cache.

> similarly, i had some standalone types.KindDatabase resources (without a db_server), and requesting with types.KindDatabase returned me no result, but the way we use the api here make it look like i should be getting some results b/c i do have dbs

Do these databases show up in the web UI?

> since this type is not in this [list](https://github.com/gravitational/teleport/blob/d2cf47f7d352edea1462baf8233fde5c4ed5643e/lib/services/unified_resource.go#L43) we use to make a lookup of access errors

This would be considered a bug probably. The clients ask for the specific type they are able to render, but we don't put any blocks around types outside of that. However, since `KindAppServer` doesn't exist in our matchers, nothing would show up anyway. I believe some documentation would solve this. But whats best is probably to just return an error if a type that the unified cache doesn't support is requested. Thoughts?

Comment (kiosion) [reply lib/web/resources.go]: > I believe some documentation would solve this. But whats best is probably to just return an error if a type that the unified cache doesn't support is requested.

I think both would be nice to have here – some more detailed comments would be helpful for explaining this.

**Comment by kiosion** [comment]
Switched the frontend to use `clusterUnifiedResourcesGet`, feels like a better implementation to me. Reverted changes to the `checkAccessToRegisteredResource` endpoint, and added TODOs for removing it plus its tests later on, as it's not used anywhere else. 

cc @flyinghermit – does this look better?

**Review by kimlisa — APPROVED** [review]
trusting that you tested the latest changes, lgtm after minor comments 👍

File: lib/web/apiserver.go
Comment (kiosion) [reply lib/web/apiserver.go:794]: Yup, just reviewed that! Will replace w/ `DELETE in 18.0` syntax, must have missed that

File: web/packages/teleport/src/services/user/user.ts
Comment (kiosion) [reply web/packages/teleport/src/services/user/user.ts:134]: it's required for the underlying `clt.ListResources` since there isn't a default set by `clusterUnifiedResourcesGet`. excluding it throws an error

**Review by kimlisa — COMMENTED** [review]
actually one more thing, we probably should update the existing api anyways, just so older UI calling the deprecated endpoint can still get the correct results.

File: lib/web/resources.go
Comment (kimlisa) [reply lib/web/resources.go]: > > similarly, i had some standalone types.KindDatabase resources (without a db_server), and requesting with types.KindDatabase returned me no result, but the way we use the api here make it look like i should be getting some results b/c i do have dbs
> 
> Do these databases show up in the web UI?

no, it doesn't show up in the web UI b/c under the hood we are really querying for `types.KindDatabaseServer` (despite requesting for `types.KindDatabase`), and I had no `types.KindDatabaseServer` resources which was why the result was empty (makes sense)

but it is confusing to be able to request for `types.KindDatabase` and get nothing in return, b/c I do have `types.KindDatabase` resources. 

i see them as two different resources and a `db` resource can exist without `db_server` resource (but a db_server cannot exist without a matching db)

It makes the [initial](https://github.com/gravitational/teleport/blob/15b345464de0f6d90a85e7a50afec91ee2b0d7f6/lib/auth/auth_with_roles.go#L1356) access checking weird too, b/c it should be more like

```
if req.Kind == types.KindDatabase, check for types.KindDatabaseServer access.
```

but we don't do that, but it's safe guarded later anyways, b/c we check for access for [each](https://github.com/gravitational/teleport/blob/15b345464de0f6d90a85e7a50afec91ee2b0d7f6/lib/auth/auth_with_roles.go#L1426) individual resource.

but i kinda understand where your coming from, like give me the `types.KindDatabase` that is being proxied by a `types.KindDatabaseServer`, but I don't feel like that it shold be used that way, because the `ListUnifiedResources` actually returns `types.KindDatabaseServer`, and then we take the result and [extract](https://github.com/gravitational/teleport/blob/15b345464de0f6d90a85e7a50afec91ee2b0d7f6/lib/web/apiserver.go#L2902) the database in the proxy

i mean at the end of the day it works, so perhaps not a big deal and I could also be just misunderstanding and making a big mess out of it :D

File: web/packages/teleport/src/services/user/user.ts
Comment (flyinghermit) [inline web/packages/teleport/src/services/user/user.ts:143]: Won't this also return true for data like `resp = {items: [null/undefined]}` ? I guess the backend wont return null here but just thinking out loud.

File: web/packages/teleport/src/services/user/user.ts
Comment (flyinghermit) [inline web/packages/teleport/src/services/user/user.ts:140]: nit: returning false would be more idiomatic here since the caller expects boolean value?

**Review by flyinghermit — APPROVED** [review]
Thank you for all the changes. 

I tested this in my cluster that only has SAML apps and now the onboarding modal is gone. I've left two comments below.

File: web/packages/teleport/src/services/user/user.ts
Comment (kiosion) [reply web/packages/teleport/src/services/user/user.ts:143]: if you mean `null`/`undefined` as entries in the list, in that case it would return true. could do something like `return !!res?.items?.some((i) => !!i)` but that feels a bit overkill to me...

File: web/packages/teleport/src/services/user/user.ts
Comment (kiosion) [reply web/packages/teleport/src/services/user/user.ts:140]: that'll return that value to `res`, the func will still always return a boolean

File: web/packages/teleport/src/services/user/user.ts
Comment (flyinghermit) [reply web/packages/teleport/src/services/user/user.ts:143]: Yes, since Go's json marshaller can send `[null]` array if there are zero items in the slice and the field is not initialized with `make()`. Though we are good here since ListUnifiedResources does not send such data. Your call how safe we want to be here ;)

**Pull request merged**
